control-plane/api/v1alpha1/controlplanerequestlimit_types.go

```diff
@@ -0,0 +1,224 @@
+	c.Status.Conditions = Conditions{
+		{
+			Type:               ConditionSynced,
+			Status:             status,
+			LastTransitionTime: metav1.Now(),
+			Reason:             reason,
+			Message:            message,
+		},
+	}
+}
+
+// SetLastSyncedTime updates the last synced time.
+func (c *ControlPlaneRequestLimit) SetLastSyncedTime(time *metav1.Time) {
+	c.Status.LastSyncedTime = time
+}
+
+// SyncedCondition gets the synced condition.
+func (c *ControlPlaneRequestLimit) SyncedCondition() (status corev1.ConditionStatus, reason, message string) {
+	cond := c.Status.GetCondition(ConditionSynced)
+	if cond == nil {
+		return corev1.ConditionUnknown, "", ""
+	}
+	return cond.Status, cond.Reason, cond.Message
+}
+
+// SyncedConditionStatus returns the status of the synced condition.
+func (c *ControlPlaneRequestLimit) SyncedConditionStatus() corev1.ConditionStatus {
+	condition := c.Status.GetCondition(ConditionSynced)
+	if condition == nil {
+		return corev1.ConditionUnknown
+	}
+	return condition.Status
+}
+
+// ToConsul converts the resource to the corresponding Consul API definition.
+// Its return type is the generic ConfigEntry but a specific config entry
+// type should be constructed e.g. ServiceConfigEntry.
+func (c *ControlPlaneRequestLimit) ToConsul(datacenter string) consul.ConfigEntry {
+	return &consul.RateLimitIPConfigEntry{
+		Kind:           c.ConsulKind(),
+		Name:           c.ConsulName(),
+		Mode:           c.Spec.Mode,
+		ReadRate:       c.Spec.ReadRate,
+		WriteRate:      c.Spec.WriteRate,
+		Meta:           meta(datacenter),
+		ACL:            c.Spec.ACL.toConsul(),
+		Catalog:        c.Spec.Catalog.toConsul(),
+		ConfigEntry:    c.Spec.ConfigEntry.toConsul(),
+		ConnectCA:      c.Spec.ConnectCA.toConsul(),
+		Coordinate:     c.Spec.Coordinate.toConsul(),
+		DiscoveryChain: c.Spec.DiscoveryChain.toConsul(),
+		Health:         c.Spec.Health.toConsul(),
+		Intention:      c.Spec.Intention.toConsul(),
+		KV:             c.Spec.KV.toConsul(),
+		Tenancy:        c.Spec.Tenancy.toConsul(),
+		PreparedQuery:  c.Spec.PreparedQuery.toConsul(),
+		Session:        c.Spec.Session.toConsul(),
+		Txn:            c.Spec.Txn.toConsul(),
+	}
+}
+
+// MatchesConsul returns true if the resource has the same fields as the Consul
+// config entry.
+func (c *ControlPlaneRequestLimit) MatchesConsul(candidate consul.ConfigEntry) bool {
+	configEntry, ok := candidate.(*consul.IngressGatewayConfigEntry)
+	if !ok {
+		return false
+	}
+	// No datacenter is passed to ToConsul as we ignore the Meta field when checking for equality.
+	return cmp.Equal(c.ToConsul(""), configEntry, cmpopts.IgnoreFields(consul.IngressGatewayConfigEntry{}, "Partition", "Namespace", "Meta", "ModifyIndex", "CreateIndex"), cmpopts.IgnoreUnexported(), cmpopts.EquateEmpty())
+}
+
+// Validate returns an error if the resource is invalid.
+func (c *ControlPlaneRequestLimit) Validate(consulMeta common.ConsulMeta) error {
+	// TODO: implement.
```

Comment:
I have no idea what to do here? Are we just validating that the inputs are within a certain range?

Reply:
@jmurret any ideas?